### PR TITLE
Add `exclude` option

### PR DESCRIPTION
**Pull request by mastrzyz**
We have a problem where we need to get a port for a local but we can't use specific ports due to them being blocked.

In our case, Chromium blocks them -> https://chromium.googlesource.com/chromium/src.git/+/refs/heads/main/net/base/port_util.cc

Here I added logic for a `exclude` generator in the same way we have the makeRange generator.

For example : 

`getPort.exclude([1024,1027])`

will give me a generator containing all ports from the valid TCP range excluding 1024 and 1027

[1025,1026,1028,1029,...,65535]

### CODE REVIEW

**Comment by mastrzyz** [comment]
@sindresorhus  let me know if I need to change anything here. Lovely tool you have in this repo!

**Comment by sindresorhus** [comment]
I think it should be an option instead. Having exclude as a separate method makes it not compose well with `makeRange()`. The `exclude` option could support any `Iterable`, so you could use `.makeRange` to make the exclusion range.

**Comment by mastrzyz** [comment]
aaa , so something like this ( current update )

**Comment by sindresorhus** [comment]
Yes, but it shouldn't even run `getAvailablePort` for the excluded ports.

**Comment by mastrzyz** [comment]
Ok I think I got it , let me know

**Comment by mastrzyz** [comment]
@sindresorhus , let me know of anything else needing to be done here, thanks!

**Comment by sindresorhus** [comment]
You haven't actually done what was asked: https://github.com/sindresorhus/get-port/pull/53#issuecomment-882916986

File: index.d.ts [hunk 1]
```diff
@@ -8,6 +8,11 @@ declare namespace getPort {
 		*/
 		readonly port?: number | Iterable<number>;
 
+		/**
+		A port or list of ports which we do not want to use
+		*/
+		readonly exclusions?: number | Array<number>;
```
Comment (sindresorhus) [inline index.d.ts:14]: It should not support a single port.

File: index.d.ts [hunk 2]
```diff
@@ -8,6 +8,11 @@ declare namespace getPort {
 		*/
 		readonly port?: number | Iterable<number>;
 
+		/**
+		A port or list of ports which we do not want to use
```
Comment (sindresorhus) [inline index.d.ts:12]: Needs a better description and should end in a `.`

File: index.js [hunk 3]
```diff
@@ -42,9 +45,16 @@ const portCheckSequence = function * (ports) {
 
 module.exports = async options => {
 	let ports;
+	let exclusions = [];
```
Comment (sindresorhus) [inline index.js:48]: For performance, it should use a `Set` internally since it's faster for checking.

File: index.js [hunk 4]
```diff
@@ -42,9 +45,16 @@ const portCheckSequence = function * (ports) {
...
 
 	if (options) {
-		ports = typeof options.port === 'number' ? [options.port] : options.port;
+		if (options.port) {
+			ports = typeof options.port === 'number' ? [options.port] : options.port;
+		}
+
+		if (options.exclusions) {
+			exclusions = typeof options.exclusions === 'number' ? [options.exclusions] : options.exclusions;
```
Comment (sindresorhus) [inline index.js:56]: There should be validation here that throws if the option is not an `Iterable<number>`.

File: readme.md [hunk 5]
```diff
@@ -62,6 +62,14 @@ Type: `number | Iterable<number>`
 
 A preferred port or an iterable of preferred ports to use.
 
+##### exclusions
```
Comment (sindresorhus) [inline readme.md:65]:
```suggestion
##### exclude
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 1 above)
Comment (mastrzyz) [reply index.d.ts]: Curious on this, why remove the flexibility?

**Comment by mastrzyz** [comment]
> You haven't actually done what was asked: [#53 (comment)](https://github.com/sindresorhus/get-port/pull/53#issuecomment-882916986)



> You haven't actually done what was asked: [#53 (comment)](https://github.com/sindresorhus/get-port/pull/53#issuecomment-882916986)

Not sure I follow, isn't that what this statement is for?

image

File: index.d.ts (hunk 1 above)
Comment (mastrzyz) [reply index.d.ts]: @sindresorhus  so I understand you correctly, we want `exclude` to only support an array of ports to exclude , not just a single port value.

**Comment by sindresorhus** [comment]
@mastrzyz I'm not entirely sure what you're asking. Please just read the comment I linked to:

> Having exclude as a separate method makes it not compose well with makeRange(). The exclude option could support **any Iterable**, so you could use .makeRange to make the exclusion range.

File: index.d.ts [hunk 6]
```diff
@@ -8,6 +8,11 @@ declare namespace getPort {
 		*/
 		readonly port?: number | Iterable<number>;
 
+		/**
+		A port or list of ports which we do not want to have found for us.
```
Comment (sindresorhus) [inline index.d.ts:12]: The doc comment is outdated and it could still be improved.

File: index.d.ts (hunk 1 above)
Comment (sindresorhus) [reply index.d.ts]: Almost. It should accept an `Iterable`.

**Comment by mastrzyz** [comment]
Closing this PR , don't have the cycles currently to continue work on this.

**Comment by mastrzyz** [comment]
I have cycles again and will start work here.

File: index.d.ts (hunk 1 above)
Comment (mastrzyz) [reply index.d.ts]: Aaa I see now, changed it to be an Iterable!

File: index.js (hunk 4 above)
Comment (mastrzyz) [reply index.js]: Done!

File: index.d.ts (hunk 6 above)
Comment (sindresorhus) [inline index.d.ts:12]:
```suggestion
		One or more ports that should not be used.
```

File: readme.md [hunk 7]
```diff
@@ -62,6 +62,14 @@ Type: `number | Iterable<number>`
 
 A preferred port or an iterable of preferred ports to use.
 
+##### exclude
+
+Type: `number | Array<number>`
```
Comment (sindresorhus) [inline readme.md:67]:
```suggestion
Type: `number[]`
```

File: readme.md [hunk 8]
```diff
@@ -62,6 +62,14 @@ Type: `number | Iterable<number>`
...
+
+A port or list of ports which we do not want to use
+
+Given an exclude of `[1024,1025]` we will never return those ports.
```
Comment (sindresorhus) [inline readme.md:71]: I don't think this sentence is needed.

File: index.test-d.ts [hunk 9]
```diff
@@ -3,6 +3,8 @@ import getPort = require('.');
 
 expectType<Promise<number>>(getPort());
 expectType<Promise<number>>(getPort({port: 3000}));
+expectType<Promise<number>>(getPort({exclude: 3000}));
```
Comment (sindresorhus) [inline index.test-d.ts:6]: Incorrect

File: index.js [hunk 10]
```diff
@@ -42,9 +45,20 @@ const portCheckSequence = function * (ports) {
...
-		ports = typeof options.port === 'number' ? [options.port] : options.port;
+		if (options.port) {
+			ports = typeof options.port === 'number' ? [options.port] : options.port;
+		}
+
+		if (options.exclude) {
+			if (typeof options.exclude[Symbol.iterator] !== 'function') {
+				throw new TypeError('Exclude option must be set as a valid iterator ( for example using "makeRange" )');
+			}
+
```
Comment (sindresorhus) [inline index.js:59]: It should also validate that all the `exclude` elements are numbers.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
You need to fix the merge conflict.

**Comment by sindresorhus** [comment]
There are still unresolved feedback.

**Comment by mastrzyz** [comment]
> There are still unresolved feedback comments.

Ack, looking at these now.

File: index.js (hunk 10 above)
Comment (mastrzyz) [reply index.js:97]: thanks for the feedback, I have added a loop to check through each value, let me know if there is a more elegant way.

**Comment by mastrzyz** [comment]
@sindresorhus  I have updated all unresolved comments to the best of my knowledge.

File: readme.md [hunk 11]
```diff
@@ -60,6 +60,12 @@ Type: `number | Iterable<number>`
 
 A preferred port or an iterable of preferred ports to use.
 
+##### exclude
+
+Type: `number[]`
+
+A port or list of ports which we do not want to use
```
Comment (sindresorhus) [inline readme.md:67]: This is not in sync with index.d.ts

File: index.js [hunk 12]
```diff
@@ -78,9 +81,28 @@ const portCheckSequence = function * (ports) {
...
+		if (options.exclude) {
+			const excludeIterable = options.exclude;
+
+			if (typeof excludeIterable[Symbol.iterator] !== 'function') {
+				throw new TypeError('The `exclude` option must be an iterable.');
+			}
+
+			for (const element of excludeIterable) {
+				if (typeof element !== 'number') {
```
Comment (sindresorhus) [inline index.js:99]: I would use `Number.isSafeInteger` here.

Comment (mastrzyz) [reply index.js:99]: Never knew that existed, thanks for the insight, I'll add that!

File: readme.md (hunk 11 above)
Comment (mastrzyz) [reply readme.md:67]: Done, thanks

File: index.js [hunk 13]
```diff
@@ -17,6 +17,9 @@ const lockedPorts = {
 // and a new young set for locked ports are created.
 const releaseOldLockedPortsIntervalMs = 1000 * 15;
 
+const portMin = 1024;
+const portMax = 65_535;
```
Comment (Richienb) [inline index.js:21]: These should be named `minPort` and `maxPort`, or maybe even `firstPort` and `lastPort`.

File: readme.md [hunk 14]
```diff
@@ -60,6 +60,12 @@ Type: `number | Iterable<number>`
 
 A preferred port or an iterable of preferred ports to use.
 
+##### exclude
+
+Type: `Iterable<number>`
+
+A port or list of ports which we do not want to use
```
Comment (Richienb) [inline readme.md:67]: There is no way to pass a single port without it being in an iterable

**Review by Richienb — CHANGES_REQUESTED** [review]

File: readme.md [hunk 15]
```diff
@@ -60,6 +60,12 @@ Type: `number | Iterable<number>`
 
 A preferred port or an iterable of preferred ports to use.
 
+##### exclude
+
+Type: `Iterable<number>`
+
+An iterable of ports which we do not want to use
```
Comment (sindresorhus) [inline readme.md:67]: The docs here should be in sync with the TS docs.

File: test.js [hunk 16]
```diff
@@ -139,6 +139,28 @@ test('makeRange produces valid ranges', t => {
 	t.deepEqual([...portNumbers(1024, 1027)], [1024, 1025, 1026, 1027]);
 });
 
+test('exclude produces ranges that exclude provided exclude list', async t => {
+	const exclude = [1024, 1026];
+	const foundPort = await getPort({exclude, port: portNumbers(1024, 1026)});
+	// We should not find any of the exclusions in the foundPorts
```
Comment (sindresorhus) [inline test.js:145]: Type "foundPorts"

**Comment by sindresorhus** [comment]
I would recommend looking over your own diff to ensure there are no more typos, inconsistencies or things that could be improved: https://github.com/sindresorhus/get-port/pull/53/files

File: readme.md (hunk 15 above)
Comment (mastrzyz) [reply readme.md:67]: Can you help me understand where are the TS Docs? are you referring to the .d.ts file?

Comment (Richienb) [reply readme.md:67]: Yes he is

**Pull request merged**